fix httpclient5.x connection leak

JoeCqupt opened the pull request.

spencergibb commented:
Hmm, locally they pass. Rerunning build.

spencergibb commented:
Any idea how to test this? The test in the reproducer in #3696 doesn't pass with this fix.

JoeCqupt commented:
> Any idea how to test this? The test in the reproducer in #3696 doesn't pass with this fix.

the reproducer code with some error config . the route uri not match mock server . 

i pass the reproducer code test by debug with evaluate code :
<img width="1920" alt="image" src="https://github.com/user-attachments/assets/87033dc3-4939-4fa0-86b4-4b7b6b7cfd9a" />

JoeCqupt commented:
> Hmm, locally they pass. Rerunning build.

the CI build behavior is strange. it still failed after rebuild

JoeCqupt commented:
> Any idea how to test this? The test in the reproducer in #3696 doesn't pass with this fix.

the reproducer code with some error config . the route uri not match mock server .

JoeCqupt commented:
@spencergibb @ryanjbaxter  PTAL

ryanjbaxter commented:
Could we use the reproducer as a basis for a test?

JoeCqupt commented:
> Could we use the reproducer as a basis for a test?

i have use the reproducer code as a basis  add a unit test .


Let me explain the code which i modified: 

first i add  `httpclient5` maven dependency for test
```
<dependency>
    <groupId>org.apache.httpcomponents.client5</groupId>
    <artifactId>httpclient5</artifactId>
    <scope>test</scope>
</dependency>
```
but this changed the underlying implementation of `ProxyExchange`.  from `REACTOR` to `HTTP_COMPONENTS` 
so i add config to `application.yml` ：
```
spring:
  http:
    client:
      factory: REACTOR
```
After modifying the config , I modified two associated unit tests： 
`GatewayServerMvcAutoConfigurationTests#gatewayHttpClientPropertiesWork `
`GatewayServerMvcAutoConfigurationTests#settingHttpClientFactoryWorks`

Finally i add a unit test to test httpcommonts as the underlying implementation of `ProxyExchange`:
```
                @Bean
		public RouterFunction<ServerResponse> gatewayRouterFunctionsRetryWithHttpComponentsClient(
				GatewayMvcProperties properties,
				ObjectProvider<HttpHeadersFilter.RequestHttpHeadersFilter> requestHttpHeadersFilters,
				ObjectProvider<HttpHeadersFilter.ResponseHttpHeadersFilter> responseHttpHeadersFilters,
				ApplicationContext applicationContext) {

			// build httpComponents client factory
			ClientHttpRequestFactory clientHttpRequestFactory = ClientHttpRequestFactoryBuilder.httpComponents()
				.withConnectionManagerCustomizer(builder -> builder.setMaxConnTotal(2).setMaxConnPerRoute(2))
				.withDefaultRequestConfigCustomizer(
						c -> c.setConnectionRequestTimeout(Timeout.of(Duration.ofMillis(3000))))
				.build();

			// build proxyExchange use httpComponents
			RestClient.Builder restClientBuilder = RestClient.builder();
			restClientBuilder.requestFactory(clientHttpRequestFactory);
			ProxyExchange proxyExchange = new RestClientProxyExchange(restClientBuilder.build(), properties);

			// build handler function use httpComponents
			ProxyExchangeHandlerFunction function = new ProxyExchangeHandlerFunction(proxyExchange,
					requestHttpHeadersFilters, responseHttpHeadersFilters);
			function.onApplicationEvent(new ContextRefreshedEvent(applicationContext));

			// @formatter:off
			return route("testretrywithhttpcomponentsclient")
					.GET("/retrywithhttpcomponentsclient", function)
					.before(new LocalServerPortUriResolver())
					.filter(retry(3))
					.filter(setPath("/retry"))
					.filter(prefixPath("/do"))
					.build();
			// @formatter:on
		}
```

ryanjbaxter commented:
Can you elaborate on why you needed to add `httpclient5`?

JoeCqupt commented:
> Can you elaborate on why you needed to add `httpclient5`?

because the reproducer code use httpclient5 as  the underlying implementation of ProxyExchange

issue : https://github.com/spring-cloud/spring-cloud-gateway/issues/3696 
reproducer code: https://github.com/johan1a/spring-cloud-gateway-mvc-connection-test/blob/master/src/main/java/com/example/demo/HttpComponentsConfig.java#L14-L20

spencergibb reviewed (changes requested):
@JoeCqupt unfortunately, adding the Apache httpclient 5 dependency will change the underlying http client for all the tests. You would need to add a new integration test module to not pollute the classpath here https://github.com/spring-cloud/spring-cloud-gateway/tree/main/spring-cloud-gateway-integration-tests

JoeCqupt commented:
> @JoeCqupt unfortunately, adding the Apache httpclient 5 dependency will change the underlying http client for all the tests. You would need to add a new integration test module to not pollute the classpath here https://github.com/spring-cloud/spring-cloud-gateway/tree/main/spring-cloud-gateway-integration-tests

OK i will add a new integration test

ryanjbaxter commented on spring-cloud-gateway-integration-tests/httpclient/pom.xml:
```diff
@@ -0,0 +1,77 @@
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.assertj</groupId>
```
Is this needed?

JoeCqupt replied:
to be prepared for future tests

ryanjbaxter replied:
I would think you would not need to include that dependency, it should come transitively....I could be wrong though.  No harm though

ryanjbaxter commented:
Can you submit this against the 4.1.x branch?
